library/core/src/intrinsics.rs

```diff
@@ -838,8 +838,11 @@ extern "rust-intrinsic" {
 
     /// Moves a value out of scope without running drop glue.
     ///
-    /// This exists solely for [`mem::forget_unsized`]; normal `forget` uses
-    /// `ManuallyDrop` instead.
+    /// This is only strictly needed for [`mem::forget_unsized`]; normal [`mem::forget`]
+    /// compiles fine just using `ManuallyDrop` instead.
+    ///
+    /// As this does literally nothing, it's trivially const-safe.
+    #[rustc_const_stable(feature = "const_forget_intrinsic", since = "1.50")]
```

Comment:
Cc @rust-lang/wg-const-eval for the const stabilization here

Reply:
@scottmcm could you add a test making sure that this actually works at const-time? From what I can see it actually should not work since this intrinsic is not implemented... you'll probably have to copy [the Miri implementation](https://github.com/rust-lang/miri/blob/2065b52dfef3cd5a5216e65c21a056a69574bddc/src/shims/intrinsics.rs#L446).

Reply:
I [added a test](https://github.com/rust-lang/rust/pull/79989/commits/479dd4c7907fab748c334eac92a4a12be66c30fb) and it seems to run fine?  Please let me know if that's not the right way to test such a thing.

Reply:
@oli-obk any idea how this test can possibly pass without `forget` being implemented [in this file](https://github.com/rust-lang/rust/blob/master/compiler/rustc_mir/src/interpret/intrinsics.rs)? I am very confused right now.

Reply:
The intrinsic does not even seem to be called:
```
├─7ms  INFO rustc_mir::interpret::step // executing bb1
├─7ms  INFO rustc_mir::interpret::step _0 = forget_arg_and_return_4(move _1) -> bb2
├┐rustc_mir::interpret::eval_context::frame FOUR_THE_HARD_WAY
│└┐rustc_mir::interpret::eval_context::frame forget_arg_and_return_4
│┌┘rustc_mir::interpret::eval_context::frame forget_arg_and_return_4
├┘rustc_mir::interpret::eval_context::frame FOUR_THE_HARD_WAY
├┐rustc_mir::interpret::eval_context::frame FOUR_THE_HARD_WAY
│└┐rustc_mir::interpret::eval_context::frame forget_arg_and_return_4
│ ├─0ms  INFO rustc_mir::interpret::step // executing bb0
│ ├─0ms  INFO rustc_mir::interpret::step StorageLive(_2)
│ ├─0ms  INFO rustc_mir::interpret::step StorageLive(_3)
│ ├─0ms  INFO rustc_mir::interpret::step _3 = move _1
│ ├─0ms  INFO rustc_mir::interpret::step _2 = std::mem::forget::<Vec<i32>>(move _3) -> [return: bb1, unwind: bb2]
│ ├┐rustc_mir::interpret::eval_context::frame forget_arg_and_return_4
│ │└┐rustc_mir::interpret::eval_context::frame std::mem::forget::<Vec<i32>>
│ │ ├─0ms  INFO rustc_mir::interpret::step // executing bb0
│ │ ├─0ms  INFO rustc_mir::interpret::step _0 = const ()
│ │ ├─0ms  INFO rustc_mir::interpret::step return
│ │ ├─0ms  INFO rustc_mir::interpret::eval_context popping stack frame (returning from function)
│ │┌┘rustc_mir::interpret::eval_context::frame std::mem::forget::<Vec<i32>>
│ ├┘rustc_mir::interpret::eval_context::frame forget_arg_and_return_4
│ ├─0ms  INFO rustc_mir::interpret::step // executing bb1
│ ├─0ms  INFO rustc_mir::interpret::step StorageDead(_3)
│ ├─0ms  INFO rustc_mir::interpret::step StorageDead(_2)
│ ├─0ms  INFO rustc_mir::interpret::step _0 = const 4_i32
│ ├─0ms  INFO rustc_mir::interpret::step return
│ ├─0ms  INFO rustc_mir::interpret::eval_context popping stack frame (returning from function)
│┌┘rustc_mir::interpret::eval_context::frame forget_arg_and_return_4
├┘rustc_mir::interpret::eval_context::frame FOUR_THE_HARD_WAY
├─7ms  INFO rustc_mir::interpret::step // executing bb2
├─7ms  INFO rustc_mir::interpret::step StorageDead(_1)
├─7ms  INFO rustc_mir::interpret::step return
├─7ms  INFO rustc_mir::interpret::eval_context popping stack frame (returning from function)
rustc_mir::interpret::eval_context::frame FOUR_THE_HARD_WAY
```
Also very strange: the trace indicates two calls to `forget_arg_and_return_4`?

Reply:
I found this, which could explain the behavior:
https://github.com/rust-lang/rust/blob/331e74014a0a2bf18aae7f02495f97958bf9767d/compiler/rustc_mir/src/transform/lower_intrinsics.rs#L27
Now I wonder if that transformation can lead to UB being masked...

Reply:
Hm, the intrinsic didn't have a body before either, so I guess there is no UB change. The intrinsic simply does not need an implementation any more.

@scottmcm you are good then, aside from the `cfg(bootstrap)` at the test which I think is unnecessary. :)

Reply:
Turns out you actually *do* need to add the intrinsic implementation, as  @tmiasko pointed out. It is still needed when libstd is built with `-Zmir-opt-level=0`.

Reply:
It should be possible to test this by calling `core::intrinsics::forget` directly, instead of `core::mem::forget` wrapper, with `-Zmir-opt-level=0`,

Reply:
I'll add the test.  Looks like with https://github.com/rust-lang/rust/pull/80040 it won't need the intrinsic implementation any more, though, right?

Reply:
> Looks like with #80040 it won't need the intrinsic implementation any more, though, right?

Yes.